File: adbus/connection.c
/* vim: ts=4 sw=4 sts=4 et
 *
 * Copyright (c) 2009 James R. McKaskill
 *
 * Permission is hereby granted, free of charge, to any person obtaining a
 * copy of this software and associated documentation files (the "Software"),
 * to deal in the Software without restriction, including without limitation
 * the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the
 * Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 * DEALINGS IN THE SOFTWARE.
 *
 * ----------------------------------------------------------------------------
 */

#define ADBUS_LIBRARY
#include "connection.h"
#include "interface.h"
#include "message.h"

#include "dmem/vector.h"
#include "dmem/string.h"

#include <assert.h>

/** \struct adbus_ConnectionCallbacks
 *  \brief Structure to hold callbacks registered with adbus_conn_new(). 
 */

/** \var adbus_ConnectionCallbacks::release
    \brief Called when the connection is destroyed.

    Since the connection is internally ref counted via adbus_conn_ref() and
    adbus_conn_deref() it can occur after the call to adbus_conn_free().
 */

/** \var adbus_ConnectionCallbacks::send_message
 *  \brief This is a callback to send messages.
    The return value should be the number of bytes written or -1 on error (ie
    the return value of recv). The connection considers any value except the
    full message length to be an error.

    For example: 
    \code
    adbus_ssize_t SendMsg(void* user, adbus_Message* m)
    { return send(*(adbus_Socket*) user, m->data, m->size, 0);
    \endcode
 */

/** \var adbus_ConnectionCallbacks::proxy 
 *
 *
    Callback to send messages across to the connection thread.
    - user - User data associated with this callback.
    - cb - Function to call on the target thread.
    - release - Function to free the data in the cbuser argument.
    - cbuser - Data associated with the cb and release arguments.

    For example: 
 *  \code
    class ProxyEvent
    {
    public:
        ProxyEvent(adbus_Callback cb, adbus_Callback release, void* user)
        : m_Cb(cb), m_Release(release), m_User(user) {}
 
        ~ProxyEvent()
        { if (m_Release) m_Release(m_User); }
 
        void call()
        { if (m_Cb) m_Cb(m_User); }
 
        adbus_Callback  m_Cb;
        adbus_Callback  m_Release;
        void*           m_User;
    };
  
    void Proxy(void* user, adbus_Callback cb, adbus_Callback release, void* cbuser)
    {
        if (OnConnectionThread())
        {
            ProxyEvent e(cb, release, user);
            e.call();
        }
        else
        {
           ProxyEvent* e = new ProxyEvent(cb, release, user);
           PostEvent(e);
        }
    }
  
    // On the connection thread
    void ProcessEvent(ProxyEvent* e)
    {
        e->call();
        delete e;
    }
    \endcode
 */

/** \struct adbus_Connection
 *
    \brief Client message dispatcher

    The connection API can be split into four sections:

    -# Getting data in/out.
    -# Proxying messages from the connection to and from other threads.
    -# Registering application callbacks via:
        -# Binds: registers an interface on a path that can be introspected,
        have methods called on, properties set, etc.
        -# Replies: registers a callback listening for method return or error
        message.
        -# Matches: generic message matches that can be pushed through to the
        bus server. This is mostly used for matching signals.
    -# Connecting to the bus server.

    \note adbus_Connection does not handle the initial authentication with the
    bus server. This can be done using either the adbus_Auth module or a third
    party SASL library.

    \warning Unless otherwise specificed all connection functions must be
    called on the connection thread and may not be called in a callback.

 *  \sa adbus_Socket, adbus_Auth, adbus_Bind, adbus_Reply, adbus_Match

    \section in Getting Data In

    The connection itself has no idea how to get incoming data or messages.
    Instead the owner of the connection figures out how to get incoming data
    off of a socket and feeds it into the connection via adbus_conn_dispatch()
    or adbus_conn_parse().

    The simplest way of handing data off to the connection is to append data
    to an adbus_Buffer and then having the connection consume complete
    messages in the buffer via adbus_conn_parse(). For example:

    \code
    #define RECV_SIZE 64 * 1024
    int ReadSignalled()
    {
        // Read all the data available
        adbus_ssize_t read = RECV_SIZE;
        while (read == RECV_SIZE)
        {
            // adbus_buf_recvbuf() gives us a place at the end of the buffer
            // to hand off to recv where it can put the data
            uint8_t* buf = adbus_buf_recvbuf(my_buffer, RECV_SIZE);
            read = recv(my_socket, RECV_SIZE, 0);
            adbus_buf_recvd(my_buffer, RECV_SIZE, read);
        }
  
        // Hand the data off to the connection. If adbus_conn_parse returns 
        // an error, we should disconnect the socket.
        if (adbus_conn_parse(my_connection, my_buffer))
            return -2;
  
        // Propagate recv errors up
        if (read < 0)
            return -1;
  
        return 0;
    }
    \endcode

 *  \sa adbus_Socket, adbus_Buffer, adbus_conn_dispatch(), adbus_conn_parse()
 *
    \section out Getting Data Out

    Outgoing data is sent via adbus_ConnectionCallbacks::send_message set with
    adbus_conn_new(). This is required to be always set.

    For example:
    \code
    static int SendMessage(void* user, adbus_Message* msg)
    {
        adbus_Socket* s = (adbus_Socket*) user;
        return send(*s, msg->data, msg->size, 0);
    }

    adbus_Connection* CreateConnection(adbus_Socket sock)
    {
        adbus_ConnectionCallbacks cbs = {};
        cbs.send_message = &SendMessage;
        return adbus_conn_new(&cbs, (void*) &sock);
    }
    \endcode

    \sa adbus_ConnectionCallbacks::send_message, adbus_conn_new()

    \section proxy Proxying Messages To/From the Connection Thread

    In multithreaded applications, the connection parsing and dispatch will be
    on a given thread, but we often want to register objects and callbacks for
    all of the application's thread on the one thread. The way we get around this
    is to proxy all messages and requests for registrations to and from the connection
    thread.

    This proxying of messages to and fro is done via the proxy callbacks setup
    in adbus_conn_new(). See proxy, should_proxy, get_proxy, and block in
    adbus_ConnectionCallbacks.

    \sa adbus_ConnectionCallbacks


    \section registrations Binds, Matches, and Replies

    See adbus_Bind, adbus_Match, and adbus_Reply.

    \section bus Bus Server

    By default the connection will not connect to the bus server (specifically
    it does not send the Hello message), until adbus_conn_connect() is called.
    This is especially useful for multithreaded applications that want to
    create a connection, register all of their objects on the various app
    threads, and then finally connect to the bus. This then avoids race
    conditions when other applications try to call method for not yet
    registered objects.

    Once the server responds to the hello message adbus_conn_isconnected()
    will return true and adbus_conn_uniquename() will return the assigned
    name.

    \sa adbus_conn_connect(), adbus_conn_isconnected(), adbus_conn_uniquename()
 */

/** \struct adbus_Match
 *
 */


// ----------------------------------------------------------------------------
// Connection management
// ----------------------------------------------------------------------------

/** Creates a new connection.
 *  \relates adbus_Connection
 */
adbus_Connection* adbus_conn_new(adbus_ConnectionCallbacks* cb, void* user)
{
    if (ADBUS_TRACE_MEMORY) {
        adbusI_log("new connection");
    }

    adbus_Connection* c = NEW(adbus_Connection);

    c->nextSerial   = 1;
    c->connected    = 0;

    c->returnMessage    = adbus_msg_new();

    c->state = adbus_state_new();

    c->callbacks = *cb;
    c->user = user;

    c->bus = adbus_proxy_new(c->state);
    adbus_proxy_init(c->bus, c, "org.freedesktop.DBus", -1, "/org/freedesktop/DBus", -1);
    adbus_proxy_setinterface(c->bus, "org.freedesktop.DBus", -1);

    adbus_Member* m;

    c->introspectable = adbus_iface_new("org.freedesktop.DBus.Introspectable", -1);

    m = adbus_iface_addmethod(c->introspectable, "Introspect", -1);
    adbus_mbr_setmethod(m, &adbusI_introspect, NULL);
    adbus_mbr_retsig(m, "s", -1);


    c->properties = adbus_iface_new("org.freedesktop.DBus.Properties", -1);

    m = adbus_iface_addmethod(c->properties, "Get", -1);
    adbus_mbr_setmethod(m, &adbusI_getProperty, NULL);
    adbus_mbr_argsig(m, "ss", -1);
    adbus_mbr_argname(m, "interface_name", -1);
    adbus_mbr_argname(m, "property_name", -1);
    adbus_mbr_retsig(m, "v", -1);

    m = adbus_iface_addmethod(c->properties, "GetAll", -1);
    adbus_mbr_setmethod(m, &adbusI_getAllProperties, NULL);
    adbus_mbr_argsig(m, "s", -1);
    adbus_mbr_argname(m, "interface_name", -1);
    adbus_mbr_retsig(m, "a{sv}", -1);

    m = adbus_iface_addmethod(c->properties, "Set", -1);
    adbus_mbr_setmethod(m, &adbusI_setProperty, NULL);
    adbus_mbr_argsig(m, "ssv", -1);
    adbus_mbr_argname(m, "interface_name", -1);
    adbus_mbr_argname(m, "property_name", -1);
    adbus_mbr_argname(m, "value", -1);

    adbus_conn_ref(c);

    return c;
}

/** Increments the connection ref count.
 *  \relates adbus_Connection
 */
void adbus_conn_ref(adbus_Connection* connection)
{ adbus_InterlockedIncrement(&connection->ref); }

// ----------------------------------------------------------------------------

/** \def adbus_conn_free(c)
 *  \brief Decrements the connection ref count.
 *  \relates adbus_Connection
 *  
 *  Since the connection is ref counted, this may not actually free the
 *  connection.
 *
 *  \sa adbus_conn_new(), adbus_conn_ref(), adbus_conn_deref()
 */

/** Decrements the connection ref count.
 *  \relates adbus_Connection
 */
void adbus_conn_deref(adbus_Connection* c)
{
    if (c && adbus_InterlockedDecrement(&c->ref) == 0) {
        if (ADBUS_TRACE_MEMORY) {
            adbusI_log("free connection");
        }

        // Remove the service lookups first, since they do not hook up to
        // release callbacks.

        for (dh_Iter si = dh_begin(&c->services); si != dh_end(&c->services); ++si) {
            if (dh_exist(&c->services, si)) {
                adbusI_freeServiceLookup(dh_val(&c->services, si));
            }
        }
        dh_clear(ServiceLookup, &c->services);

        // Clear out the lookup tables. These will disconnect from the binds,
        // replies, and matches so that the bind etc free methods dont try and
        // disconnect from the lookup tables. 

        for (dh_Iter ii = dh_begin(&c->paths); ii != dh_end(&c->paths); ++ii) {
            if (dh_exist(&c->paths, ii)){
                struct ObjectPath* p = dh_val(&c->paths, ii);
                p->connection = NULL;
                adbusI_freeObjectPath(p);
            }
        }
        dh_clear(ObjectPath, &c->paths);

        for (dh_Iter ri = dh_begin(&c->remotes); ri != dh_end(&c->remotes); ++ri) {
            if (dh_exist(&c->remotes, ri)) {
                struct Remote* r = dh_val(&c->remotes, ri);
                r->connection = NULL;
                adbusI_freeRemote(r);
            }
        }
        dh_clear(Remote, &c->remotes);

        // Release and free the connection services.  Note we protect against
        // the release callback removing the any other services that have yet
        // to be freed

        adbus_ConnBind* b;
        DIL_FOREACH(Bind, b, &c->binds, fl) {
            adbusI_freeBind(b);
        }

        adbus_ConnReply* r;
        DIL_FOREACH(Reply, r, &c->replies, fl) {
            adbusI_freeReply(r);
        }

        adbus_ConnMatch* m;
        DIL_FOREACH(Match, m, &c->matches, hl) {
            adbusI_freeMatch(m);
        }

        // Assert that the above release callbacks have not added any more
        // services

        assert(dil_isempty(&c->binds));
        assert(dil_isempty(&c->replies));
        assert(dil_isempty(&c->matches));
        assert(dh_size(&c->services) == 0);
        assert(dh_size(&c->paths) == 0);
        assert(dh_size(&c->remotes) == 0);

        dh_free(ServiceLookup, &c->services);
        dh_free(ObjectPath, &c->paths);
        dh_free(Remote, &c->remotes);

        adbus_state_free(c->state);

        adbus_proxy_free(c->bus);
        adbus_iface_free(c->introspectable);
        adbus_iface_free(c->properties);

        adbus_msg_free(c->returnMessage);

        free(c->uniqueService);

        dv_free(char, &c->parseBuffer);

        free(c);
    }
}

// ----------------------------------------------------------------------------
// Message
// ----------------------------------------------------------------------------

/** Sends a message on the connection
 *  \relates adbus_Connection
 *  
 *  This function is thread safe and may be called in both callbacks and
 *  on other threads.
 *
 *  \note Do not call this is in method call callback for replies. Instead
 *  setup your response in the provided msg factory (adbus_CbData::ret).
 *
 *  \todo Actually make this thread safe.
 */
int adbus_conn_send(
        adbus_Connection* c,
        adbus_Message*    message)
{
    if (ADBUS_TRACE_MSG) {
        adbusI_logmsg("sending", message);
    }

    assert(message->serial != 0);

    if (!c->callbacks.send_message)
        return -1;

    adbus_ssize_t sent = c->callbacks.send_message(c->user, message);
    if (sent != (adbus_ssize_t) message->size)
        return -1;

    return 0;
}

// ----------------------------------------------------------------------------

/** Gets a serial that can be used for sending messages.
 *  \relates adbus_Connection
 *  
 *  \note This function is thread safe and may be called in both callbacks and
 *  on other threads.
 *
 *  \todo Actually make this thread safe.
 */
uint32_t adbus_conn_serial(adbus_Connection* c)
{
    if (c->nextSerial == UINT32_MAX)
        c->nextSerial = 1;
    return c->nextSerial++;
}





// ----------------------------------------------------------------------------
// Parsing and dispatch
// ----------------------------------------------------------------------------

static int Dispatch(adbus_CbData* d)
{
    if (d->msg->size == 0)
        return 0;

    if (ADBUS_TRACE_MSG) {
        adbusI_logmsg("received", d->msg);
    }

    if (    d->msg->type == ADBUS_MSG_METHOD
        &&  !d->msg->flags & ADBUS_MSG_NO_REPLY
        &&  d->ret) {
        adbus_msg_reset(d->ret);
    } else {
        d->ret = NULL;
    }

    // Dispatch the method call
    if (d->msg->type == ADBUS_MSG_METHOD)
    {
        if (adbusI_dispatchBind(d))
            return -1;

        // Setup the needed return fields
        if (d->ret) {
            if (adbus_msg_type(d->ret) == ADBUS_MSG_INVALID)
                adbus_msg_settype(d->ret, ADBUS_MSG_RETURN);

            if (!adbus_msg_serial(d->ret))
                adbus_msg_setserial(d->ret, adbus_conn_serial(d->connection));

            adbus_msg_setflags(d->ret, adbus_msg_flags(d->ret) | ADBUS_MSG_NO_REPLY);

            adbus_msg_setreply(d->ret, d->msg->serial);
            if (d->msg->sender)
                adbus_msg_setdestination(d->ret, d->msg->sender, d->msg->senderSize);

        }
    }

    // Dispatch match and reply
    adbus_MsgFactory* ret = d->ret;
    d->ret = NULL;
    int err = adbusI_dispatchMatch(d) || adbusI_dispatchReply(d);
    d->ret = ret;

    return err;
}

// ----------------------------------------------------------------------------

/** Dispatch a pre-parsed message
 *  \relates adbus_Connection
 */
int adbus_conn_dispatch(
        adbus_Connection* c,
        adbus_Message*    message)
{
    if (message->size == 0)
        return 0;

    adbus_CbData d;
    memset(&d, 0, sizeof(adbus_CbData));
    d.connection    = c;
    d.msg           = message;
    d.ret           = c->returnMessage;

    if (Dispatch(&d))
        return -1;

    // Send off reply if needed
    if (d.ret) {
        adbus_msg_send(d.ret, c);
    }

    return 0;
}

// ----------------------------------------------------------------------------

/** Consume messages from the supplied buffer.
 *  \relates adbus_Connection
 *
 *  This will remove all complete messages from the beginning of the buffer,
 *  but it will leave incomplete messages in the buffer. These should be
 *  appended to once more data comes in and then recall this function.
 */
int adbus_conn_parse(
        adbus_Connection*   c,
        adbus_Buffer*      buf)

{
    char* data = adbus_buf_data(buf);
    size_t size = adbus_buf_size(buf);

    adbus_Message m;

    while (1) {
        size_t msgsize = adbus_parse_size(data, size);
        if (msgsize == 0 || msgsize > size)
            break;

        if (ADBUS_ALIGN(data, 8) == (uintptr_t) data) {
            if (adbus_parse(&m, data, msgsize))
                return -1;
            if (adbus_conn_dispatch(c, &m))
                return -1;

        } else {
            char* dest = dv_push(char, &c->parseBuffer, msgsize);
            memcpy(dest, data, msgsize);
            if (adbus_parse(&m, dest, msgsize))
                return -1;
            if (adbus_conn_dispatch(c, &m))
                return -1;
            dv_clear(char, &c->parseBuffer);
        }

        data += msgsize;
        size -= msgsize;
    }

    adbus_buf_remove(buf, 0, adbus_buf_size(buf) - size);
    return 0;
}

// ----------------------------------------------------------------------------

/** See if calling code should use adbus_conn_proxy()
 *  \relates adbus_Connection
 *  \sa adbus_ConnectionCallbacks::should_proxy
 */
adbus_Bool adbus_conn_shouldproxy(adbus_Connection* c)
{ return c->callbacks.should_proxy && c->callbacks.should_proxy(c->user) && c->callbacks.proxy; }

// ----------------------------------------------------------------------------

/** Proxy a call over to the connection thread or call immediately if already
 * on it.
 *  \relates adbus_Connection
 *
 * \sa adbus_conn_shouldproxy(), adbus_ConnectionCallbacks::proxy
 */
void adbus_conn_proxy(adbus_Connection* c, adbus_Callback callback, void* user)
{ 
    if (c->callbacks.proxy) {
        c->callbacks.proxy(c->user, callback, user); 
    } else {
        callback(user);
    }
}

// ----------------------------------------------------------------------------

/** Get proxy functions for the current thread.
 *  \relates adbus_Connection
 *
 *  The arguments may be null if that value is not needed.
 *
 *  \sa adbus_ConnectionCallbacks::get_proxy
 */
void adbus_conn_getproxy(
        adbus_Connection*       c,
        adbus_ProxyCallback*    cb,
        adbus_ProxyMsgCallback* msgcb,
        void**                  user)
{
    if (c->callbacks.get_proxy)
    {
        c->callbacks.get_proxy(c->user, cb, msgcb, user);
    }
    else
    {
        if (cb)
            *cb = NULL;
        if (msgcb)
            *msgcb = NULL;
        if (user)
            *user = NULL;
    }
}

File: include/lua/adbus/connection.lua
-------------------------------------------------------------------------------
-- vim: ts=4 sw=4 sts=4 et tw=78
--
-- Copyright (c) 2009 James R. McKaskill
--
-- Permission is hereby granted, free of charge, to any person obtaining a
-- copy of this software and associated documentation files (the "Software"),
-- to deal in the Software without restriction, including without limitation
-- the rights to use, copy, modify, merge, publish, distribute, sublicense,
-- and/or sell copies of the Software, and to permit persons to whom the
-- Software is furnished to do so, subject to the following conditions:
--
-- The above copyright notice and this permission notice shall be included in
-- all copies or substantial portions of the Software.
--
-- THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
-- IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
-- FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
-- AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
-- LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
-- FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
-- DEALINGS IN THE SOFTWARE.
-------------------------------------------------------------------------------

smodule 'adbus'
require 'adbuslua_core'
require 'pretty'

-------------------------------------------------------------------------------
-- Connection API
-------------------------------------------------------------------------------

local mt = {}
mt.__index = mt

_M.connect_address = adbuslua_core.connect_address
_M.bind_address = adbuslua_core.bind_address

DEBUG = false

function _M.enable_debug(dbg)
    if dbg == nil or dbg then
        DEBUG = true
        adbuslua_core.set_log_level(2)
    else
        DEBUG = false
        adbuslua_core.set_log_level(0)
    end
end

--- Creates a connection
-- \param bus       The bus to connect to.
-- \param debug     boolean flag to enable/disable debugging (default
--                  disabled)
--
-- \return The new connection object
--
-- The bus can be one of:
-- - nil - we then connect to the default bus ('session')
-- - 'session' or 'system' to connect to the session or system buses
--   respectively
-- - A string giving the bus address in dbus form eg.
--   'tcp:host=localhost,port=12345'

local connections = {}

function _M.connection(bus)
    local self = connections[bus or 'default']
    if self then return self end

    local self = setmetatable({}, mt)

    self._queue             = {}
    self._connection        = adbuslua_core.connection.new(bus, self._queue)
    self._current_message   = nil
    self._return_message    = nil
    self._return_yield      = {}

    self._callback = function()
        self._return_message = self:current_message()
        self._return_yield.yield = true
    end

    connections[bus or 'default'] = self

    return self
end

function mt:close()
    self._connection:close()
end

--- Returns whether we successfully connected to the bus.
function mt:is_connected_to_bus()
    return self._connection:is_connected_to_bus()
end

--- Returns our unique name if we are connected to the bus or nil.
function mt:unique_name()
    return self._connection:unique_name()
end

--- Returns a new serial for use with messages.
function mt:serial()
    return self._connection:serial()
end

function mt:process(yieldtable)
    while not yieldtable.yield do

        -- Process the queue
        local i = 1
        while true do
            local msg = self._queue[i]
            self._queue[i] = nil

            if not msg then 
                break 
            end

            i = i + 1

            if DEBUG then
                print(table.show(msg, "Received"))
            end

            self._current_message = msg

            if msg.callback then

                -- Match or reply callback
                if msg.type == "error" then
                    msg.callback(nil, msg.error, msg[1])
                else
                    msg.callback(unpack(msg))
                end

            elseif msg.object then
                
                -- Method call
                local obj = msg.object
                local retsigs = msg.user
                local method = msg.member

                if msg.no_reply then
                    obj[method](obj, unpack(msg))
                else
                    local ret = {obj[method](obj, unpack(msg))}
                    ret.destination = msg.sender
                    ret.reply_serial = msg.serial
                    if ret[1] == nil and ret[2] ~= nil then
                        -- 1 is nil
                        -- 2 is error name
                        -- 3 is error msg
                        ret[1] = ret[3]
                        ret.error_name = ret[2]
                        ret.type = "error"
                        ret.signature = ret[1] and "s" or ""
                    else
                        ret.type = "method_return"
                        ret.signature = retsigs[method]
                    end
                    self:send(ret)
                end

            end

            self._current_message = nil

            if yieldtable.yield then
                return
            end
        end

        -- Get more data
        self._connection:process()
    end
end

function mt:current_message()
    return self._current_message
end

--- Adds a match rule to the connection
--
-- The param is a table detailing the fields of messages to match against.
-- Only the callback is required. Optional fields not specified will always
-- match.
--
-- Arguments:
-- \param m.callback            Callback for when the match is hit
-- \param[opt] m.type           type of message. Valid values are:
--                               - 'method_call'
--                               - 'method_return'
--                               - 'error'
--                               - 'signal'
-- \param[opt] m.sender         Service name
-- \param[opt] m.destination    Service name
-- \param[opt] m.interface      String
-- \param[opt] m.reply_serial   Number
-- \param[opt] m.path           String
-- \param[opt] m.member         String
-- \param[opt] m.error          String
-- \param[opt] m.remove_on_first_match      Boolean
-- \param[opt] m.add_to_bus     Boolean
-- \param[opt] m.object         Anything
-- \param[opt] m.id             Number - should be an id returned from
--                              connection:match_id.
-- \param[opt] m.unpack_message Boolean
-- \return The match id. This will be the value of m.id if it is set.
--
-- Details:
-- The sender and destination fields should be valid dbus names (ascii dot
-- seperated string). They can either be a unique name (begins with : eg
-- :1.34) or a service name (reverse dns eg org.freedesktop.DBus) in which
-- case the service will be tracked.
--
-- The path should be a valid dbus path name (unix style path eg
-- /org/freedesktop/DBus).
--
-- The id can be filled out with a return number from connection:match_id
-- or it can be left blank in which case an id will be generated. Either way
-- the id is returned and should be used as the paramater to remove_match.
--
-- By default the match will be kept active until the match is explicitely
-- removed or the connection is destroyed. Alternatively setting
-- m.remove_on_first_match will remove the match the first time it hits a
-- match. This is most often used to match return values.
--
-- The match rule will normally be a local only match and thus can match
-- method_calls (method_calls should really be redirected via an interface),
-- method_returns and errors, but in order to match a signal the match rule
-- must be added to the bus daemon by setting m.add_to_bus.
-- Otherwise the bus will not redirect the signal message our way. This is not
-- necessary for non bus connection.
--
-- The arguments to the callback depend on two params (unpack_message and
-- object). If object is set then the first argument to the function will be
-- that object. If unpack_message is set to true or not set (ie default), the
-- message arguments will be unpacked so the call to the callback will be of
-- the form:
--
-- callback(object, arg0, arg1, ...) or callback(arg0, arg1, ...)
--
-- unpacked error message take a special form. In this case the callback is
-- callback is called with the object (optional), a nil, the message error
-- name, and then the message description (if provided). eg:
--
-- callback(object, nil, 'org.freedesktop.DBus.Error.UnknownMethod',
--          'org.freedesktop.DBus does not understand message foo')
--
-- If unpack_message is set to false then the second argument (if object is
-- set) will be a message object. See connection.send for the format of 
-- this object. eg
-- 
-- callback(object, message) or callback(message)
--
-- Example:
-- \code
-- local function owner_changed(name, from, to)
-- end
--
-- c:add_match{
--      type = 'signal',
--      add_to_bus = true,
--      sender = 'org.freedesktop.DBus',
--      member = NameOwnerChanged
-- }
-- \endcode
--          
--
-- \sa  connection:remove_match
--      connection:send_messsage
--
function mt:add_match(m)
    if DEBUG then
        print(table.show(m, "Adding match"))
    end

    return self._connection:add_match(m.callback, m)
end

function mt:add_reply(remote, serial, callback)
    if DEBUG then
        print(table.show(m, "Adding reply"))
    end

    return self._connection:add_reply(remote, serial, callback)
end

local bind = {}
bind.__index = bind

function bind:emit(member, ...)
    local sig = self._interface._signal_signature[member]
    if not sig then
        error("Invalid signal name")
    end

    local msg = {
        type        = 'signal',
        path        = self._path,
        interface   = self._interface.name,
        signature   = sig,
        member      = member,
        ...
    }

    self._connection:send(msg)
end

function mt:bind(object, path, interface)
    if type(interface) == "string" then
        interface = _M.interface.new(interface)
    end

    local user      = interface._method_return_signature
    local b         = setmetatable({}, bind)
    b._bind         = self._connection:bind(object, user, path, interface._interface)
    b._interface    = interface
    b._path         = path
    b._connection   = self
    return b
end


--- Sends a new message
-- 
-- If arguments are provided, then the signature must be set. The format of
-- the signature can be either:
-- - A single string of the dbus signatures of each argument concatanated
--   together
-- - A table of strings of dbus signatures for each argument
--
-- The dbus signature is an ascii string that represents the marshalled type.
-- Valid values for this are:
--
-- The simple types are:
-- y    - 8 bit unsigned integer
-- b    - Boolean value
-- n    - 16 bit signed integer
-- q    - 16 bit unsigned integer
-- i    - 32 bit signed integer
-- u    - 32 bit unsigned integer
-- x    - 64 bit signed integer
-- t    - 64 bit unsigned integer
-- d    - 64 bit floating point number (double)
-- s    - string
-- o    - dbus object path
-- g    - dbus signature
-- 
-- The compound types are (using X as a placeholder for the inner type):
-- aX    - an array of values of type X
--       - X must be a single conceptual type
-- (X)   - a structure of values of type X 
--       - X can be one or more types in sequence
-- v     - a variant where each occurance dictates the type with the data
-- a{XY} - a dictionary/table/map with keys of type X and values of type Y
--
-- The marshalling from lua values can be grouped into a number of categories:
-- - numeric dbus values (8, 16, 32, 64 bit integers and double) expect a lua
--   number argument
-- - boolean dbus values expect a lua boolean argument
-- - string dbus values (string, object path, and signature) expect a lua
--   string
-- - dbus arrays expect a table of the appropriate values of length #table
--   using integer indices starting at 1 (standard lua arrays)
-- - dbus maps expect a table with the appropriate key and value types
-- - dbus variants will marshall in a number of ways
--   - a lua number is marshalled as a double
--   - a lua boolean is marshalled as a boolean
--   - a lua string is marshalled as a string
--   - a lua table is marshalled as an array of variants if table[1] ~= nil 
--     or as a dbus map with keys and values as variants
--   - if the value
--
--
-- The argument is a message object. The same format is also used for packed
-- match callbacks. This is a table with a number of fields:
-- \param m.type                The type of message. This can be one of:
--                               - 'method_call'
--                               - 'method_return'
--                               - 'error'
--                               - 'signal'
-- \param[opt] m.serial         The serial of the message. If set this should
--                              be set using a return of connection.next_serial.
-- \param[opt] m.interface      Interface field
-- \param[opt] m.path           Path field
-- \param[opt] m.member         Member field
-- \param[opt] m.error          error name field
-- \param[opt] m.destination    Destination field
-- \param[opt] m.sender         Sender field
-- \param[opt] m.signature      Signature of the message. See above.
-- \param[opt] m.no_reply       If set to true, sets the NO_REPLY_EXPECTED
--                              flag in the message. This is an optimisation
--                              where the remote side can then choose to not
--                              send back a reply.
-- \param[opt] m.no_auto_start  If set to true, sets the NO_AUTO_START flag in
--                              the message. This is an indication to the bus
--                              to not auto start a daemon to handle the
--                              message.
-- \param[opt] m[1], m[2] ...   The arguments of the message. They are
--                              marshalled using the signature field.
--
-- \sa  connection.add_match 
--      http://dbus.freedesktop.org/doc/dbus-specification.html
--
function mt:send(m)
    if DEBUG then
        print(table.show(m, "Sending"))
    end

    self._connection:send(m)
end

-- Method callers

function mt:async_call(msg)
    local serial = self:serial()

    local reply
    if msg.callback then
        reply = self:add_reply(msg.destination, serial, msg.callback)
    end

    msg.type = "method_call"
    msg.serial = serial
    msg.callback = nil

    self:send(msg)

    return reply
end

function mt:call(msg)
    if not msg.no_reply then
        msg.callback = self._callback
    end

    local reply = self:async_call(msg)
    msg._callback = nil

    if reply then
        self._return_yield.yield = false
        self._return_message = nil
        self:process(self._return_yield)
        
        local ret = self._return_message
        if ret.type == "error" then
            return nil, ret.error, ret[1]
        else
            return unpack(ret)
        end
    end
end


